File: app/src/main/java/vn/ngh/example/ListUserAdapter.java
package vn.ngh.example;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.List;

import vn.ngh.stickyitem.R;
import vn.ngh.stickyitem.interfaces.StickyAdapter;

/**
 * Created by Ngh on 7/13/2017.
 */

public class ListUserAdapter extends RecyclerView.Adapter<ListUserAdapter.ContentViewHolder>
        implements StickyAdapter<ListUserAdapter.StickyViewHolder> {

    private int mCurrentUserIndex = RecyclerView.NO_POSITION;
    private List<User> mUserList;
    private LayoutInflater mInflater;
    private Context mContext;
    private boolean isUpdate = true;

    static class ContentViewHolder extends RecyclerView.ViewHolder {
        private TextView mTextName;
        private TextView mTextEmail;
        private ImageView mImageAvatar;

        public ContentViewHolder(View itemView) {
            super(itemView);
            mTextName = (TextView) itemView.findViewById(R.id.text_name);
            mTextEmail = (TextView) itemView.findViewById(R.id.text_email);
            mImageAvatar = (ImageView) itemView.findViewById(R.id.image_avatar);
        }
    }

    static class StickyViewHolder extends RecyclerView.ViewHolder {
        private TextView mTextName;
        private TextView mTextEmail;
        private ImageView mImageAvatar;

        public StickyViewHolder(View itemView) {
            super(itemView);
            mTextName = (TextView) itemView.findViewById(R.id.text_name);
            mTextEmail = (TextView) itemView.findViewById(R.id.text_email);
            mImageAvatar = (ImageView) itemView.findViewById(R.id.image_avatar);
        }
    }

    public ListUserAdapter(Context context, List<User> userList) {
        this(context, userList, RecyclerView.NO_POSITION);
    }

    public ListUserAdapter(Context context, List<User> userList, int currentUserIndex) {
        this.mUserList = userList;
        this.mContext = context;
        this.mInflater = LayoutInflater.from(mContext);
        this.mCurrentUserIndex = currentUserIndex;
        if (mCurrentUserIndex == 0 && mUserList.size() == 1)
            this.mCurrentUserIndex = RecyclerView.NO_POSITION;
    }

    @Override
    public ContentViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = mInflater.inflate(R.layout.item_user, parent, false);
        ContentViewHolder viewHolder = new ContentViewHolder(view);
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(ContentViewHolder holder, int position) {
        if (holder != null) {
            int stickyIndex = getStickyIndex();
            int pos = position >= stickyIndex && stickyIndex != RecyclerView.NO_POSITION
                    ? position + 1 : position;
            User user = mUserList.get(pos);
            holder.mTextName.setText(user.getName());
            holder.mTextEmail.setText(user.getEmail());
            holder.mImageAvatar.setImageResource(user.getIdAvatar());
        }
    }

    @Override
    public StickyViewHolder onCreateStickyViewHolder(ViewGroup parent) {
        View view = mInflater.inflate(R.layout.item_user, parent, false);
        StickyViewHolder viewHolder = new StickyViewHolder(view);
        return viewHolder;
    }

    @Override
    public void onBindStickyViewHolder(StickyViewHolder viewHolder, int position) {
        if (mCurrentUserIndex >= 0 && viewHolder != null) {
            User user = mUserList.get(mCurrentUserIndex);
            viewHolder.mTextName.setText(user.getName());
            viewHolder.mTextEmail.setText(user.getEmail());
            viewHolder.mImageAvatar.setImageResource(user.getIdAvatar());
        }
    }

    @Override
    public int getItemCount() {
        if (mUserList != null) {
            if (mCurrentUserIndex >= 0 && mUserList.size() > 1) return mUserList.size() - 1;
            else return mUserList.size();
        } else return 0;
    }

    @Override
    public int getStickyIndex() {
        return mCurrentUserIndex;
    }

    @Override
    public int getDataCount() {
        return getItemCount();
    }

    @Override
    public boolean getEnableUpdate() {
        return isUpdate;
    }

    @Override
    public void setEnableUpdate(boolean enableUpdate) {
        isUpdate = enableUpdate;
    }

    public int getCurrentUserIndex() {
        return mCurrentUserIndex;
    }

    public void setCurrentUserIndex(int mCurrentUserIndex) {
        this.mCurrentUserIndex = mCurrentUserIndex;
    }
}

File: app/src/main/java/vn/ngh/example/ListUserAdapter.kt
package vn.ngh.example

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.item_sticky.view.*
import kotlinx.android.synthetic.main.item_user.view.*
import vn.ngh.stickyitem.R
import vn.ngh.stickyitem.interfaces.StickyAdapter

/**
 * Created by Ngh
 */

class ListUserAdapter(mContext: Context, private val mUserList: List<User>?, currentUserIndex: Int = RecyclerView.NO_POSITION) : RecyclerView.Adapter<ListUserAdapter.ContentViewHolder>(), StickyAdapter<ListUserAdapter.StickyViewHolder> {

    private var currentUserIndex = RecyclerView.NO_POSITION
    private val mInflater: LayoutInflater = LayoutInflater.from(mContext)
    private var isUpdate = true

    class ContentViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

    class StickyViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

    init {
        this.currentUserIndex = currentUserIndex
        if (this.currentUserIndex == 0 && mUserList?.size == 1)
            this.currentUserIndex = RecyclerView.NO_POSITION
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ContentViewHolder {
        val view = mInflater.inflate(R.layout.item_user, parent, false)
        return ContentViewHolder(view)
    }

    override fun onBindViewHolder(holder: ContentViewHolder?, position: Int) {
        if (holder != null) {
            val stickyIndex = getStickyIndex()
            val pos = if (position >= stickyIndex && stickyIndex != RecyclerView.NO_POSITION)
                position + 1
            else
                position
            mUserList?.let {
                val user = it[pos]
                holder.itemView.textName.text = user.name
                holder.itemView.textEmail.text = user.email
                holder.itemView.textAvatar.text = position.toString()
            }
        }
    }

    override fun onCreateStickyViewHolder(parent: ViewGroup): StickyViewHolder {
        val view = mInflater.inflate(R.layout.item_sticky, parent, false)
        return StickyViewHolder(view)
    }

    override fun onBindStickyViewHolder(viewHolder: StickyViewHolder, position: Int) {
        if (currentUserIndex >= 0) {
            mUserList?.let {
                val user = it[currentUserIndex]
                viewHolder.itemView.textNameSticky.text = user.name
                viewHolder.itemView.textEmailSticky.text = user.email
                viewHolder.itemView.imageAvatar.setImageResource(user.idAvatar)
            }
        }
    }

    override fun getItemCount(): Int {
        return if (mUserList != null) {
            if (currentUserIndex >= 0 && mUserList.size > 1)
                mUserList.size - 1
            else
                mUserList.size
        } else
            0
    }

    override fun getStickyIndex(): Int {
        return currentUserIndex
    }

    override fun getDataCount(): Int {
        return itemCount
    }

    override fun getEnableUpdate(): Boolean {
        return isUpdate
    }

    override fun setEnableUpdate(enableUpdate: Boolean) {
        isUpdate = enableUpdate
    }
}
